[web] Encode AssetManifest.bin as JSON and use that on the web.

ditman opened the pull request:
This PR modifies the web build slightly to create an `AssetManifest.json`, that is a JSON(base64)-encoded version of the `AssetManifest.bin` file.

_(This should enable all browsers to download the file without any interference, and all servers to serve it with the correct headers.)_

It also modifies Flutter's `AssetManifest` class so it loads and uses said file `if (kIsWeb)`.

### Issues

* Fixes https://github.com/flutter/flutter/issues/124883

### Tests

* Unit tests added.
* Some tests that run on the Web needed to be informed of the new filename, but their behavior didn't have to change (binary contents are the same across all platforms).
* I've deployed a test app, so users affected by the BIN issue may take a look at the PR in action:
  * https://dit-tests.web.app


## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

ditman committed:
[web] Encode AssetManifest.bin as JSON and use that on the web.

jonahwilliams commented:
base64 decoding is pretty slow. Have you benchmarked this versus the new binary for and/or the old JSON format?

ditman commented:
> Have you benchmarked this versus the new binary for and/or the old JSON format?

@jonahwilliams Nope, do those benchmarks exist or do they need to be written?

yjbanov commented on packages/flutter/lib/src/services/asset_manifest.dart:
```diff
@@ -21,6 +23,14 @@ abstract class AssetManifest {
+        // Decode the manifest JSON file to the underlying BIN, and convert to ByteData.
+        final ByteData message = ByteData.sublistView(base64.decode(json.decode(jsonData) as String));
+        // Now we can keep operating as usual.
+        return _AssetManifestBin.fromStandardMessageCodecMessage(message);
```
As a quick sanity check, can you use `FlutterTimeline` or something to time how long it takes to decode `jsonData` into an instance of `_AssetManifestBin`?

ditman replied:
Definitely. I've also redeployed the app in the description in `--profile` mode so we can look at the flame chart from chrome.

jonahwilliams commented:
See https://github.com/flutter/flutter/blob/master/dev/benchmarks/microbenchmarks/lib/foundation/decode_and_parse_asset_manifest.dart

jonahwilliams commented:
This change will also still require us to make an additional network requst before any assets can be downloaded.

ditman commented:
@jonahwilliams @yjbanov

I took the microbenchmark that Jonah suggested [here](https://github.com/flutter/flutter/pull/131382#issuecomment-1652876205) (thanks!), and modified it slightly to run on the web.

I had to:

* Add a `web/index.html` (I actually did `flutter create --platform web .`)
* Changed the number of iterations from 1000 to 10000 (with 1k, each time I refreshed the benchmark, `master` would be even faster, so I assumed this was the browser warming up. I raised the number of iterations to try to give a chance to all benchmarks to be "similarly warm" without having to run them several times).

I ran the benchmark in `master` and the 2 PRs. I ran each example like this:

```
refresh-tools && flutter clean && flutter run -d web-server --target lib/foundation/decode_and_parse_asset_manifest.dart --web-port 7357 --release
```

(`refresh-tools` is a small command that I have to delete the `flutter_tools` cache, to ensure that changes actually work)

Here's the results after 10k iterations:

| master | #130419 | #131382 (this) |
|--------|---------|----------------|
| <img width="786" alt="master" src="https://github.com/flutter/flutter/assets/1255594/fc1dbf57-fcc2-4ae6-91c4-1de6886fd0d3"> | <img width="787" alt="130" src="https://github.com/flutter/flutter/assets/1255594/026438f7-70c4-435e-8c5c-6b5056f641b1"> | <img width="786" alt="131" src="https://github.com/flutter/flutter/assets/1255594/8634001d-7945-41b7-b431-778193265639"> |
| 57078ms | 58278ms | 59956ms |

I only ran each microbenchmark once in my gLinux machine, so I take the numbers with a healthy amount of skepticism. Squinting a little bit, we could say that "this" is 2878ms slower than `master`, so I guess downloading a ~30% larger file and json-base64-decoding it incurs in an (average(, not the best for benchmarks)) ~0.3ms penalty.

It also seems that #130419 is still downloading the `AssetManifest.bin` on the web, which is hurting it badly in this benchmark, I expected it to wipe the floor with the other two branches (I don't know exactly if I messed up here, but it seems that the global window array is being set, so changes are being applied. I can also take a look there and see what's up.)

jonahwilliams commented:
FYI @andrewkolos maybe you can help out with the benchmark issue here? Do we need to use a different method to load and parse the manifest?

yjbanov commented:
Thanks for benchmarking this, @ditman! Looks like there's no material performance difference between these approaches. If so, I'd say let's go with the simplest solution.

ditman commented:
@yjbanov Andrew's PR should be much, much faster, I'm shaving that yak now to find why is it downloading the manifest (when it's set in the DOM already)

ditman commented:
I pushed a [fix](https://github.com/flutter/flutter/pull/130419/commits/dd6ce29b8387cced79df0508284811857bcc489c) to @andrewkolos' PR, and NOW it's chugging!

<img width="555" alt="Screenshot 2023-07-27 at 3 46 13 PM" src="https://github.com/flutter/flutter/assets/1255594/b190d0a2-bb34-475a-9159-70105994ec9a">

 11ms for 10k iterations of parsing the manifest (now that the bytes are getting inlined, there's not much work to do for the loader, other than juggling the bytes coming from the JS global).

/cc @jonahwilliams @yjbanov.

ditman committed:
Make change more backwards compatible.

* Leave old AssetManifest.json, as it's public API for now.
* Encode AssetManifest.bin as AssetManifest.bin.json for the web.
* Ensure only AssetManifest.bin OR AssetManifest.bin.json are created.
* Happy case unit tests.

ditman committed:
Use the correct file, and appease analyzer.

ditman committed:
Continue generating AssetManifest.bin even for web, it is used by other tests.

jonahwilliams commented:
What are the numbers here? Sorry its a bit hard to follow the different issues/threads.

andrewkolos commented:
Heya, sorry for the silence--I was out at the tail end of last week.

I initially considered generating a separate version of the asset manifest for web, but I wanted to cut out the network request for the asset manifest to speed up (initial) asset loading and avoid having all pending asset loads depending on a single network request). However, if we are willing to delay addressing these properly (i.e. fixing #118115), I don't mind this adopting this solution. The other reason I avoided a solution like this one is that I wanted to avoid having to write/maintain a bunch of logic to construct/consume a JSON file, but your solution of just taking the existing bin data and encoding it as a string is pretty clever and doesn't strike me as _too_ slow or data-consuming.

Barring strong opinions from others, I'd like to adopt this PR and close my original PR (after a few days or so).

ditman commented:
Thanks @andrewkolos! I'll take a second look at the tests I broke here and get this one ready to land. I'd like to give this version a shot before we do drastic changes to the build process, but when we do those, we'll be sure to include both the font manifests and the assets one!

ditman committed:
Tell TestAssetBundle about AssetManifest.bin.json

ditman commented:
Agh, I was forgetting to pass `--platform=chrome` to my `flutter test` commands, that's why things were working locally but still failing in CI!

ditman committed:
Do not use isolates to loadString on the web.

ditman committed:
Emit the proper AssetManifest.bin.json in tests.

ditman commented:
@andrewkolos offered to prepare a g3Fix for this so the "Google Testing" check passes and the PR can land. (It may be *this* PR that one that lands, or some other that @andrewkolos owns so he can do any changes he needs!)

andrewkolos commented:
One challenge I failed to foresee with this is the amount of change that might be required in g3. Unlike past asset manifest changes, this adds a file that is _conditionally_ included in the build. This makes updating tests and build rules more difficult. In addition (though not as important), these kind of changes tend to break tests for customers that mock/fake `AssetBundle` in their tests. One of the reasons I initially liked this PR over my initial PR was it's relative simplicity. However, when considering g3 and testing, it may not be so simple.

It's possible that updating g3 to work with this PR might not be _that_ complicated, but there are unknowns. Given that I've already invested at least 3x as much time as I wanted to on this issue, I'd like to go with the option that will probably have the fewest surprises. As a result, I may revisit my previous PR (https://github.com/flutter/flutter/pull/130419).

ditman commented:
> It's possible that updating g3 to work with this PR might not be that complicated, but there are unknowns.

OK, I guess I'll give that a shot too.

andrewkolos commented:
> > It's possible that updating g3 to work with this PR might not be that complicated, but there are unknowns.
> 
> ~OK, I guess I'll give that a shot too.~
> 
> @andrewkolos is right, this is a world of hurt internally, what a shame :/

Do you have opinions against the core idea of https://github.com/flutter/flutter/pull/130419? As far as I can tell, @jonahwilliams and @yjbanov are satisfied with the core idea, as long as it uses a public dart API (as opposed to a hidden JS global). I really don't want to foist a bunch of g3 work on you. I'm happy to rework #130419 instead (though I can only hope g3 doesn't use custom code for generating the web entry point 🙃--I never got to find out because FroB bugged out and refused to test the PR).

ditman commented:
I have a g3fix ready for this, applying.

ditman commented:
> I can only hope g3 doesn't use custom code for generating the web entry point 🙃

@andrewkolos sorry, I've just seen this message 😅

In fact, g3 does uses custom code to generate the web entrypoint (see `cs//depot/google3/mobile/flutter/build_defs/internal/web_rules.bzl;l=195`), so you'd still have to go through the g3fix process to land the other version of the change.

I prefer this PR because it keeps about the same complexity that we currently have in `master`, fixes the problem that some people have, and gives us some extra time to actually design what's going to be our web inlining strategy. (As Jonah said, it's not the responsibility of your PR to find that solution.)

I guess I'm trying to avoid having one-offs for each file we may want to inline, and find a consolidated solution. What if tomorrow we want to make the FontManifest.json a binary file? (I'd rather it became a .bin.json for initial web solution, than creating another different setter to the engine, etc...)

ditman commented:
Google testing should pass with this g3fix: cl/553547021

christopherfujino commented:
@ditman looks like the testing cl didn't apply your g3fix: cl/553348554. Should we retry?

christopherfujino commented:
friendly ping @ditman

ditman commented:
(Apologies, I've been OOO for the last few weeks!)

@christopherfujino Yes please!

This PR has supposedly passed a TGP run (cl/553273631), but I never got it to run alongside its g3fix (cl/553547021).

If there's anything you can do (or that I need to do) to move things along, I'd love to get it done!

christopherfujino commented:
> (Apologies, I've been OOO for the last few weeks!)
> 
> @christopherfujino Yes please!
> 
> This PR has supposedly passed a TGP run (cl/553273631), but I never got it to run alongside its g3fix (cl/553547021).
> 
> If there's anything you can do (or that I need to do) to move things along, I'd love to get it done!

Well unrelated, Google testing is borked: https://github.com/flutter/flutter/issues/133814

ditman commented:
😿

ditman commented:
Looks like the FROB test passed, yay! Can you PTAL @yjbanov?

yjbanov commented on packages/flutter/lib/src/services/asset_manifest.dart:
```diff
@@ -2,25 +2,37 @@
   /// Loads asset manifest data from an [AssetBundle] object and creates an
   /// [AssetManifest] object from that data.
   static Future<AssetManifest> loadFromAssetBundle(AssetBundle bundle) {
+    if (kIsWeb) {
```
nit: a couple of lines (or links) above the `if` explaining how/why things are different on the web would be useful to future contributors.

ditman replied:
(Done)

ditman commented:
Working on addressing yegor's nits (also giving some internal code time to land before landing this fix!)

ditman committed:
Adding some comments to loadFromAssetBundle.

ditman commented:
Adding `autosubmit`, 🤞

fhuonder commented:
Hi @ditman 

Do you know when this fix will land on a stable release?
We currently have the problem that in some companies the firewalls block the bin-file of our flutter-web-app.

Regards,
Florian
